### PR TITLE
Fix query and path params not keeping the coerced types

### PR DESCRIPTION
**Pull request by Fetchinator7**
This is to fix requests not saving with the correct types.
~~NOTE: This is only for query parameters, to my understanding path parameters are not being serialized at all. See line 307 in src/router.ts~~
~~typescript
{
  name: queryParam,
  in: 'query',
},~~

Update: I realized coercing query parameters and path parameters use the exact same logic so I changed this to coerce both types.
#144

### CODE REVIEW

File: src/router.ts [hunk 1]
```diff
@@ -325,6 +327,26 @@ export class OpenAPIRouter<D extends Document = Document> {
               const commaParsed = parseQuery(commaQueryString, { comma: true });
               query[queryParam] = commaParsed[queryParam];
             }
+            const schemaType = (parameter.schema as PickVersionElement<D, OpenAPIV3.SchemaObject, OpenAPIV3_1.SchemaObject>)?.type;
+            if (Array.isArray(query[queryParam])) {
+              query[queryParam] = (query[queryParam] as any[]).map((par) => {
+                // coerce the input to the specified type
+                if (schemaType === 'boolean') {
+                  return par === 'false' ? false : Boolean(par);
```
Comment (anttiviljami) [inline src/router.ts:335]: I'm thinking we need a more robust way to parse booleans from strings.

```
Boolean('0') // false
```

File: src/router.ts [hunk 2]
```diff
@@ -325,6 +327,26 @@ export class OpenAPIRouter<D extends Document = Document> {
               const commaParsed = parseQuery(commaQueryString, { comma: true });
               query[queryParam] = commaParsed[queryParam];
             }
+            const schemaType = (parameter.schema as PickVersionElement<D, OpenAPIV3.SchemaObject, OpenAPIV3_1.SchemaObject>)?.type;
+            if (Array.isArray(query[queryParam])) {
+              query[queryParam] = (query[queryParam] as any[]).map((par) => {
```
Comment (anttiviljami) [inline src/router.ts:332]: Not a big fan of the `any[]` here.

Maybe let's do something like:

```
type ParamValue = string | number | boolean

export interface ParsedRequest extends Request {
  params: {
    [key: string]: ParamValue | ParamValue[];
  };
  cookies: {
    [key: string]: ParamValue | ParamValue[];
  };
  query: {
    [key: string]: ParamValue | ParamValue[];
  };
  requestBody: any;
}
```

Then I believe we don't need to cast the type here at all. `Array.isArray` should be a sufficent Type Guard

**Comment by Fetchinator7** [comment]
[I'm thinking we need a more robust way to parse booleans from strings](https://github.com/anttiviljami/openapi-backend/pull/262#discussion_r767178040)
Ok, I guess my question is; "If the input parameter string is not a boolean, how should we handle that in the parser?" For example, `parseJSON(input.toLowerCase())` would return the appropriate boolean or undefined if the input is anything besides those strings, but if the input isn't a boolean what would we do? Would we throw an error here in the parser? From my understanding, validating request inputs happens elsewhere with ajv (or whatever is implemented) and this is only for parsing the requests.
So, if the input is invalid/not a boolean is it the responsibility of this parser to throw an error? This works correctly for the happy cases, so what should happen in the unhappy cases that would allow us to implement "a more robust way to parse booleans"?

File: src/router.ts (hunk 1 above)
Comment (Fetchinator7) [reply src/router.ts]: I didn't know about [ajv type coercion](https://ajv.js.org/coercion.html) when I first did this but I found out about it so I changed this to use that instead.

**Comment by nklisch** [comment]
@anttiviljami Do we expect a fix, like this PR is suppose to address, likely to happen? Does this PR need to be updated? As I see it is a year old?

**Comment by anttiviljami** [comment]
Superceded by #571

**Pull request closed without merging**